### ISSUE TITLE
CUDA-accelerated BA / sparse solvers

### ISSUE THREAD

**Issue by pwais** [colmap/colmap#2643]
**Is your feature request related to a problem? Please describe.**
An early experiment showed that there was large speed-up using _dense_ solvers https://github.com/colmap/colmap/pull/2161

The goal is to leverage GPU to speed up BA solving in general.

**Describe the solution you'd like**
Recently Ceres merged cuDSS support for sparse Cholesky https://github.com/ceres-solver/ceres-solver/commit/6fb3dae4eeef855568e47ebbb29a8ba4f3c9153f   and the unit tests show an example of CUDA sparse Bundle Adjustment: https://github.com/ceres-solver/ceres-solver/commit/6fb3dae4eeef855568e47ebbb29a8ba4f3c9153f#diff-dc637bf6d41e3df3a877146387833e8147c045b9b0476a1c55b0f3f0ffc12a0fR49 

Perhaps these new changes could be adopted into COLMAP

**Additional context**
Another PR aiming at faster BA: https://github.com/colmap/colmap/pull/2567

**Comment by sarlinpe** [comment]
I agree that this would be a valuable addition, feel free to give it a try if you have time. We could start from https://github.com/colmap/colmap/pull/2161 and support the sparse solvers as well (which will be available only in the next version of Ceres). I would automatically fall back to the CPU version is the installed Ceres version is not sufficiently high.

**Comment by ahojnnes** [comment]
Thank you @dkargin. There is also quite a bit of literature on the topic, including the previously in colmap supported PBA backend: https://grail.cs.washington.edu/projects/mcba/pba.pdf

It could be interesting to see the speedup of just plugging in the new ceres sparse cuda solver and evaluate if and what speedup we could gain with such a simple change without any other customization.

**Comment by whuaegeanse** [comment]
> 1. All calculations must be done on GPU. GPU-CPU transfers are very expensive compared to calculation speed. Both linear solver (PCG, Schur+PCG, ILU+PCG, whatever) and evaluation of residuals and jacobians and hessian stuff must be done exclusively on GPU.

For very large problems, such as more than 10,000 images, is the memory on GPU large enough?

> 2. After everything is done on GPU, you will find that optimization time of dataset like BAL/final/problem-4585-1324582-pre.txt can be done in around **3-5 seconds** on RTX 2080. Ceres needs about **180 seconds** with single thread to solve that (about 30x performance boost). But the time you need just to build optimization problem (create all factors and add it to OptimizationProblem) could take about **10 seconds** alone. There is std::vectors in each CostFunction, which makes registration of the problem much slower. And building internal layout for jacobian storage is relatively expensive as well. So generally, you must refactor the way the problem is registered and the way residuals/jacobians are evaluated. Ceres/gtsam/g2o approaches are not very compatible with high performance GPU evaluation.

The speed  of GPU accelerated BA is really impressive.
Evaluate residuals/jacobians on the GPU side? Compared with the Ceres solver,  the solver will become less flexible.

> 3. You will need to rebuild all the architecture again for multi-GPU solver.

What does “all the architecture“ mean?

**Comment by dkargin** [comment]
>For very large problems, such as more than 10,000 images, is the memory on GPU large enough?
BAL/final/problem-13682-4456117-pre.txt (13682 images) barely fits RTX2080Ti with PCG-based solver and **float** as main data type. 
Large scale datasets use PCG or implicit Schur PCG (Ceres::ITERATIVE_SCHUR). Jacobians are major consumers of VRAM. Neither Hessian nor Schur complement matrices are explicitly built in such solvers.
A100 will easily handle more than 10,000 images. Multi-GPU project was mostly intended to overcome VRAM limitation, especially when you do not have A100.

>Evaluate residuals/jacobians on the GPU side? Compared with the Ceres solver, the solver will become less flexible.
We used a serious trickery with unified memory for user data (observations) to support evaluation of Ceres-like factors on GPU. And used some black magic to map host-side virtual tables to GPU-side virtual tables. It was very fragile, but worked. 
You can still build general-purpose solver for GPU, but with larger evaluation blocks. Like "this shader/kernel computes all reprojection errors of this particular type and geometry". This approach will work much faster.

>What does “all the architecture“ mean?
Architecture describes the way you store internal layout of jacobian matrix, hessian matrix or preconditioner stuff. In case of multi-GPU you have multiple isolated layouts for every GPU for every part of the solver, and additional data intersections. And then everything becomes even more tricky with mandatory async operations for calculations and exchange or multi-gpu scalability will suffer.
